src/dbnode/storage/index.go

```diff
@@ -838,8 +838,45 @@ func (i *nsIndex) writeBatchForBlockStart(
 	// Note: attemptTotal should = attemptSkip + attemptWrite.
 	i.metrics.asyncInsertAttemptWrite.Inc(int64(numPending))
 
-	// i.e. we have the block and the inserts, perform the writes.
-	result, err := i.activeBlock.WriteBatch(batch)
+	// Split the batch into warm and cold so that warm can go to the active block
+	// while cold can go to the associated cold time block. This is important to prevent
+	// a case where cold index data has not been flushed to disk yet, but gets cleaned up from
+	// the active block. We ensure that can't happen by always writing the cold index in-mem data
+	// directly.
+	warmBatch := index.NewWriteBatch(batch.Options())
+	coldBatch := index.NewWriteBatch(batch.Options())
```

Comment:
This is the simplest implementation for splitting. Assuming this approach works I can fix this up to avoid needless allocs (e.g. if no cold writes, just use the original batch instead of newing two more, etc.)

Reply:
Since there's only ever a single goroutine at a time ever calling `writeBatchForBlockStart(...)` can we actually just use a lock (like a new "writeBatchesLock") and then always reuse the same warm batches and cold batches struct? (not sure if it's reuseable but should technically be trivial to make reuseable).